File: java/src/main/java/com/myriad/wdl/model/definitions/WdlWorkflow.java
package com.myriad.wdl.model.definitions;

import com.myriad.wdl.model.WdlDocument.WdlDocumentElement;
import com.myriad.wdl.model.base.WdlNode;
import java.util.ArrayDeque;
import lombok.Getter;
import lombok.Setter;

/**
 * Workflow definition node.
 *
 * <p>In the WDL specification, a workflow composes calls, declarations, scatters, and conditionals
 * into a larger pipeline. This class preserves workflow body order so that processors can model
 * scoping and after-dependencies correctly.
 */
public final class WdlWorkflow implements WdlDocumentElement {

  /** Marker for any node that can appear directly inside a workflow body. */
  public interface WdlWorkflowElement extends WdlNode {}

  @Getter @Setter private String name;
  private final ArrayDeque<WdlWorkflowElement> elements = new ArrayDeque<>();

  public WdlWorkflow() {}

  public WdlWorkflow(String name) {
    this.name = name;
  }

  /** Returns the ordered workflow body elements exactly as they appeared in the source. */
  public ArrayDeque<WdlWorkflowElement> getElements() {
    return elements;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }
}

File: python/src/wdl_model/model/definitions/wdl_workflow.py
"""Workflow definition nodes for the Python WDL object model."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.base import WdlNode


class WdlWorkflowElement(WdlNode):
    """Marker for any node that can appear directly inside a workflow body."""

    pass


@dataclass
class WdlWorkflow:
    """Workflow definition node.

    Workflows compose calls, declarations, scatters, and conditionals into a larger pipeline.
    """

    name: str | None = None
    _elements: deque[WdlWorkflowElement] = field(default_factory=deque)

    def getElements(self) -> deque[WdlWorkflowElement]:
        """Return the ordered workflow body elements exactly as they appeared in the source."""
        return self._elements

    def elements(self) -> deque[WdlWorkflowElement]:
        """Alias for ``getElements`` for Python-facing callers."""
        return self._elements
